### PR TITLE
CLDC-1731 Allow data coordinators to view parent's schemes

### PR DESCRIPTION
**Pull request by kosiakkatrina**
Display parent schemes for data coordinators:
Do not allow editing, deactivating schemes and locations or adding new locations to parent organisation scheme.

### CODE REVIEW

File: app/controllers/locations_controller.rb [hunk 1]
```diff
@@ -225,11 +225,15 @@ def authenticate_scope!
...
 
   def authenticate_action!
-    if %w[create update index new_deactivation deactivate_confirm deactivate postcode local_authority name units type_of_unit mobility_standards availability check_answers].include?(action_name) && !((current_user.organisation == @scheme&.owning_organisation) || current_user.support?)
+    if %w[create update index new_deactivation deactivate_confirm deactivate postcode local_authority name units type_of_unit mobility_standards availability check_answers].include?(action_name) && !user_allowed_action?
       render_not_found and return
     end
   end
 
+  def user_allowed_action?
+    (current_user.organisation == @scheme&.owning_organisation) || (current_user.organisation.parent_organisations.any? { |org| org == @scheme&.owning_organisation }) || current_user.support?
```
Comment (ghost) [inline app/controllers/locations_controller.rb]: you'd probably make this check a bit more efficient by moving `current_user.support?` to the front

File: app/controllers/schemes_controller.rb [hunk 2]
```diff
@@ -265,11 +265,15 @@ def find_resource
...
     head :unauthorized and return unless current_user.data_coordinator? || current_user.support?
 
-    if %w[show locations primary_client_group confirm_secondary_client_group secondary_client_group support details check_answers edit_name deactivate].include?(action_name) && !((current_user.organisation == @scheme&.owning_organisation) || current_user.support?)
+    if %w[show locations primary_client_group confirm_secondary_client_group secondary_client_group support details check_answers edit_name deactivate].include?(action_name) && !user_allowed_action?
       render_not_found and return
     end
   end
 
+  def user_allowed_action?
+    (current_user.organisation == @scheme&.owning_organisation) || (current_user.organisation.parent_organisations.any? { |org| org == @scheme&.owning_organisation }) || current_user.support?
```
Comment (ghost) [inline app/controllers/schemes_controller.rb]: same comment as above

File: app/controllers/locations_controller.rb (hunk 1 above)
Comment (ghost) [reply app/controllers/locations_controller.rb]:
```
(current_user.organisation.parent_organisations.any? { |org| org == @scheme&.owning_organisation })
```
Could use [exists?](https://apidock.com/rails/ActiveRecord/Base/exists%3F/class) and run this check with a single query without having to store anything in memory

File: app/controllers/locations_controller.rb [hunk 3]
```diff
@@ -225,11 +225,15 @@ def authenticate_scope!
   end
 
   def authenticate_action!
-    if %w[create update index new_deactivation deactivate_confirm deactivate postcode local_authority name units type_of_unit mobility_standards availability check_answers].include?(action_name) && !((current_user.organisation == @scheme&.owning_organisation) || current_user.support?)
+    if %w[create update index new_deactivation deactivate_confirm deactivate postcode local_authority name units type_of_unit mobility_standards availability check_answers].include?(action_name) && !user_allowed_action?
```
Comment (ghost) [inline app/controllers/locations_controller.rb]: can't this be moved to a more standard `except` `only` to the top of the file?

File: app/helpers/locations_helper.rb [hunk 4]
```diff
@@ -84,6 +84,10 @@ def location_creation_success_notice(location)
     end
   end
 
+  def user_can_edit_scheme?(user, scheme)
+    user.support? || user.organisation == scheme.owning_organisation
```
Comment (ghost) [inline app/helpers/locations_helper.rb:88]: could move this check to a policy,

File: app/controllers/locations_controller.rb (hunk 1 above)
Comment (ghost) [reply app/controllers/locations_controller.rb]: and could also move this logic to a Policy

File: app/controllers/schemes_controller.rb (hunk 2 above)
Comment (kosiakkatrina) [reply app/controllers/schemes_controller.rb]: Nice!

File: app/controllers/locations_controller.rb (hunk 1 above)
Comment (kosiakkatrina) [reply app/controllers/locations_controller.rb]: Yup, can do!

File: app/controllers/locations_controller.rb [hunk 5]
```diff
@@ -225,13 +226,13 @@ def authenticate_scope!
   end
 
   def authenticate_action!
-    if %w[create update index new_deactivation deactivate_confirm deactivate postcode local_authority name units type_of_unit mobility_standards availability check_answers].include?(action_name) && !user_allowed_action?
+    unless user_allowed_action?
       render_not_found and return
```
Comment (ghost) [inline app/controllers/locations_controller.rb]: you might not need the return anymore here but not 100% sure

File: app/controllers/locations_controller.rb [hunk 6]
```diff
@@ -225,13 +226,13 @@ def authenticate_scope!
...
     end
   end
 
   def user_allowed_action?
-    (current_user.organisation == @scheme&.owning_organisation) || (current_user.organisation.parent_organisations.any? { |org| org == @scheme&.owning_organisation }) || current_user.support?
+    current_user.support? || current_user.organisation == @scheme&.owning_organisation || current_user.organisation.parent_organisations.exists?(@scheme&.owning_organisation_id)
```
Comment (ghost) [inline app/controllers/locations_controller.rb:235]: can this happen?

`current_user.organisation.parent_organisations.exists?(nil) == true` ?

Comment (kosiakkatrina) [reply app/controllers/locations_controller.rb:235]: `Organisation.exists?(nil)` and `Organisation.exists?(nil&.id)` both return false, so should be fine

File: app/helpers/locations_helper.rb (hunk 4 above)
Comment (kosiakkatrina) [reply app/helpers/locations_helper.rb:88]: Hmm, don't know if this alone merits a policy 🤔

**Review by ghost — APPROVED** [review]

**Comment by asmega** [comment]
I feel like this will somewhat clash with my PR https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1642

**Pull request merged**